### PR TITLE
fix(workflow): use $sourceDescriptions for workflow references

### PR DESCRIPTION
**Pull request by harshit078**
## What/Why/How?

- Added support for the Arazzo workflow reference form `$sourceDescriptions.<name>.<workflowId>`. 
- The previous `$sourceDescriptions.<name>.workflows.<workflowId>` form is still maintained.

## Reference

https://github.com/Redocly/redocly-cli/issues/2826

## Check yourself

- [x] This PR follows the [contributing guide](https://github.com/Redocly/redocly-cli/blob/main/CONTRIBUTING.md#pull-request-guidelines) 
- [x] All new/updated code is covered by tests
- [ ] Core code changed? - Tested with other Redocly products (internal contributions only)
- [ ] New package installed? - Tested in different environments (browser/node)
- [x] Documentation update has been considered 

## Security

- [x] The security impact of the change has been considered 
- [x] Code follows company security practices and guidelines

---

> [!NOTE]
> **Low Risk**
> Backward-compatible context parsing change with broad test updates; no auth or security surface touched.
> 
> **Overview**
> Adds **Arazzo-compliant** external workflow references as `$sourceDescriptions.<name>.<workflowId>` (without a `workflows` segment) in `getValueFromContext`, so Respect can resolve referenced workflows the way the spec describes.
> 
> The older `$sourceDescriptions.<name>.workflows.<workflowId>` form still works. For the spec form, resolution looks up the workflow in `workflows` first; if there is no match, it can return another property on the source description (e.g. `url`). Legacy paths that do not resolve still return `undefined` as before.
> 
> Tests and flow-runner examples were updated to use the spec form, with extra coverage for legacy compatibility and spec edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9608bc5edde825c6393513f399495259ce851cbf. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by harshit078** [comment]
Hi @DmitryAnansky , Can you review the PR and check it ? Thanks !

File: packages/respect-core/src/modules/context-parser/get-value-from-context.ts [hunk 1]
```diff
@@ -316,26 +316,43 @@ const resolveValue = (
...
+        (workflow: Workflow) => workflow.workflowId === workflowId
+      );
 
-    return sourceDescriptions[sourceDescriptionName].workflows.find(
-      (workflow: Workflow) => workflow.workflowId === workflowId
-    );
+      if (workflow) {
+        return workflow;
+      }
+    }
```
Comment (DmitryAnansky) [inline packages/respect-core/src/modules/context-parser/get-value-from-context.ts:342]: Probably we need to handle case with 
```
    if (isLegacyForm) {
      return undefined;
    }
```

File: packages/respect-core/src/modules/context-parser/get-value-from-context.ts [hunk 2]
```diff
@@ -316,26 +316,43 @@ const resolveValue = (
...
     const parts = path.split('.');
 
     const sourceDescriptionName = parts[1];
-    const workflowId = parts[3];
+    const isLegacyForm = parts.length === 4 && parts[2] === 'workflows';
 
-    if (!sourceDescriptionName || !workflowId) {
-      return undefined;
-    }
+    if (isLegacyForm) {
```
Comment (DmitryAnansky) [inline packages/respect-core/src/modules/context-parser/get-value-from-context.ts]: Seems we can simplify the code below(326-344) with just
```
const workflowId = isLegacyForm ? parts[3] : parts.length === 3 ? parts[2] : undefined;
```

**Comment by DmitryAnansky** [comment]
Could you please update existing tests to use correct and not legacy format?

**Comment by harshit078** [comment]
Hi @DmitryAnansky, I have addressed the comments left by you and resolved cursor bot comments. Can you review the changes again ? thanks !

File: packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts [hunk 3]
```diff
@@ -208,6 +208,98 @@ describe('getValueFromContext', () => {
...
+      workflowId: 'workflowTestId',
+      steps: [],
+    });
+  });
+
+  it('should fall back to field access for the spec form when no workflow matches', () => {
+    const ctx = {
+      $sourceDescriptions: {
+        test: {
+          url: './test.yaml',
```
Comment (DmitryAnansky) [inline packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts:240]: Can there actually be a url  on the same level as workflows?
```
$sourceDescriptions: bundledDescriptions,
```
Please check create-test-context.ts:44-63.

Comment (DmitryAnansky) [reply packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts:240]: Looks like synthetic context structure due to not strict typing.

File: packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts [hunk 4]
```diff
@@ -143,7 +143,7 @@ describe('getValueFromContext', () => {
   });
 
   // $sourceDescriptions.<name>.workflows.<workflowId>
```
Comment (DmitryAnansky) [inline packages/respect-core/src/modules/__tests__/config-parser/get-value-from-context.test.ts:145]:
```suggestion
// $sourceDescriptions.<name>.<workflowId> (Arazzo spec form)
// $sourceDescriptions.<name>.workflows.<workflowId> (legacy form)
```

File: packages/respect-core/src/modules/context-parser/get-value-from-context.ts [hunk 5]
```diff
@@ -316,26 +316,33 @@ const resolveValue = (
     return path.slice(7, -2);
   }
 
+  // $sourceDescriptions.<name>.<workflowId>
   // $sourceDescriptions.<name>.workflows.<workflowId>
```
Comment (DmitryAnansky) [inline packages/respect-core/src/modules/context-parser/get-value-from-context.ts]:
```suggestion
// $sourceDescriptions.<name>.<workflowId> (Arazzo spec form)
// $sourceDescriptions.<name>.workflows.<workflowId> (legacy form)
```

**Comment by DmitryAnansky** [comment]
@harshit078 what do you think about slightly different implementation that could cover workflow references resolution and not silently quit with unefined.
We could possibly create a separate file and move our helper logic for `parseSourceDescriptionWorkflowRef` and 
`resolveWorkflowReference`:

packages/respect-core/src/modules/context-parser/resolve-workflow-reference.ts
```
import { type RuntimeExpressionContext, type TestContext, type Workflow } from '../../types.js';

// $sourceDescriptions.<name>.<workflowId> (Arazzo spec form)
// $sourceDescriptions.<name>.workflows.<workflowId> (legacy form)
export function parseSourceDescriptionWorkflowRef(
  ref: string
): { sourceDescriptionName: string; workflowId: string; isLegacyForm: boolean } | undefined {
  if (!ref.startsWith('$sourceDescriptions.')) {
    return undefined;
  }

  const parts = ref.split('.');

  if (parts.length === 3 && parts[1] && parts[2]) {
    return { sourceDescriptionName: parts[1], workflowId: parts[2], isLegacyForm: false };
  }

  if (parts.length === 4 && parts[2] === 'workflows' && parts[1] && parts[3]) {
    return { sourceDescriptionName: parts[1], workflowId: parts[3], isLegacyForm: true };
  }

  return undefined;
}

export function resolveWorkflowReference({
  ref,
  ctx,
}: {
  ref: string | undefined;
  ctx: TestContext | RuntimeExpressionContext;
}): Workflow | undefined {
  if (!ref) {
    return undefined;
  }

  if (!ref.startsWith('$')) {
    return 'workflows' in ctx
      ? ctx.workflows?.find((workflow) => workflow.workflowId === ref)
      : undefined;
  }

  const parsedRef = parseSourceDescriptionWorkflowRef(ref);

  if (!parsedRef) {
    return undefined;
  }

  const workflows = ctx.$sourceDescriptions?.[parsedRef.sourceDescriptionName]?.workflows;

  return Array.isArray(workflows)
    ? workflows.find((workflow: Workflow) => workflow.workflowId === parsedRef.workflowId)
    : undefined;
}
```
-----

then we could update our `get-value-from-context` condition to not return undefined, but throw errors when we have not valid refs

```
if (path.startsWith('$sourceDescriptions.')) {
    const parsedRef = parseSourceDescriptionWorkflowRef(path);

    if (parsedRef) {
      const { sourceDescriptionName, workflowId: reference, isLegacyForm } = parsedRef;
      const sourceDescriptions = getFrom(ctx)('$sourceDescriptions');
      const sourceDescriptionDocument = sourceDescriptions?.[sourceDescriptionName];

      if (!sourceDescriptionDocument) {
        throw new Error(
          `Can't resolve ${red(path)}: source description ${red(
            sourceDescriptionName
          )} is not found. Available source descriptions: ${Object.keys(
            sourceDescriptions ?? {}
          ).join(', ')}.`
        );
      }

      const workflow = resolveWorkflowReference({ ref: path, ctx });

      if (workflow) {
        return workflow;
      }

      if (!isLegacyForm) {
        // Raw Source Description Objects (name, url, type) are only carried by the
        // test context — deliberately not exposed on the runtime expression context.
        const sourceDescriptionObject = (
          'sourceDescriptions' in ctx ? ctx.sourceDescriptions : undefined
        )?.find(({ name }) => name === sourceDescriptionName);

        if (
          sourceDescriptionObject &&
          Object.prototype.hasOwnProperty.call(sourceDescriptionObject, reference)
        ) {
          return (sourceDescriptionObject as Record<string, unknown>)[reference];
        }

        if (Object.prototype.hasOwnProperty.call(sourceDescriptionDocument, reference)) {
          return sourceDescriptionDocument[reference];
        }
      }

      const availableWorkflows = (
        Array.isArray(sourceDescriptionDocument.workflows)
          ? (sourceDescriptionDocument.workflows as Workflow[])
          : []
      )
        .map((workflow) => workflow.workflowId)
        .join(', ');
      const availableWorkflowsHint = availableWorkflows
        ? ` Available workflows: ${availableWorkflows}.`
        : '';

      throw new Error(
        isLegacyForm
          ? `Can't resolve ${red(path)}: workflow ${red(
              reference
            )} is not found in source description ${red(
              sourceDescriptionName
            )}.${availableWorkflowsHint}`
          : `Can't resolve ${red(path)}: ${red(
              reference
            )} does not match a workflow or a field of source description ${red(
              sourceDescriptionName
            )}.${availableWorkflowsHint}`
      );
    }
    // Other shapes ($sourceDescriptions.<name>, deep paths) resolve generically below.
  }
````

This will also require additional changes in the execution flow.
run-step.ts
```
....
    const targetWorkflow = resolveWorkflowReference({ ref: targetWorkflowRef, ctx });
....
        // register the step in catch sections so the failed check is displayed and counted in the totals
        ctx.executedSteps.push(step);
        printUnknownStep(step, ctx.options.logger);
.....

and changes to the `runner.ts` as we need to register failed workflow when dependsOn not resolved properly:
```
....
try {
        await handleDependsOn({ workflow, ctx, config: options.config, executedStepsCount });
      } catch (error) {
        if (!(error instanceof WorkflowDependencyNotFoundError)) {
          throw error;
        }

        // an unresolvable dependsOn reference fails this workflow but must not
        // abort the remaining workflows and files
        const startTime = performance.now();
        const failedStep = {
          stepId: 'dependsOn',
          checks: [
            {
              name: CHECKS.UNEXPECTED_ERROR,
              message: error.message,
              passed: false,
              severity: ctx.severity['UNEXPECTED_ERROR'],
            },
          ],
        } as Step;

        printWorkflowSeparator({
          fileName: basename(ctx.options.filePath),
          workflowName: workflow.workflowId,
          logger: options.logger,
        });
        printUnknownStep(failedStep, options.logger);

        executedWorkflows.push({
          type: 'workflow',
          workflowId: workflow.workflowId,
          startTime,
          endTime: performance.now(),
          totalTimeMs: performance.now() - startTime,
          executedSteps: [failedStep],
          ctx,
          globalTimeoutError: false,
        });
        continue;
      }
    }
    ```
    
    
@harshit078 Could you please review this idea.
Its raw implementation and will require additional test coverage, but might help to handle missing workflow referenses.

**Comment by DmitryAnansky** [comment]
> @harshit078 what do you think about slightly different implementation that could cover workflow references resolution and not silently quit with unefined. We could possibly create a separate file and move our helper logic for `parseSourceDescriptionWorkflowRef` and `resolveWorkflowReference`:
> 
> packages/respect-core/src/modules/context-parser/resolve-workflow-reference.ts
> 
> ```
> import { type RuntimeExpressionContext, type TestContext, type Workflow } from '../../types.js';
> 
> // $sourceDescriptions.<name>.<workflowId> (Arazzo spec form)
> // $sourceDescriptions.<name>.workflows.<workflowId> (legacy form)
> export function parseSourceDescriptionWorkflowRef(
>   ref: string
> ): { sourceDescriptionName: string; workflowId: string; isLegacyForm: boolean } | undefined {
>   if (!ref.startsWith('$sourceDescriptions.')) {
>     return undefined;
>   }
> 
>   const parts = ref.split('.');
> 
>   if (parts.length === 3 && parts[1] && parts[2]) {
>     return { sourceDescriptionName: parts[1], workflowId: parts[2], isLegacyForm: false };
>   }
> 
>   if (parts.length === 4 && parts[2] === 'workflows' && parts[1] && parts[3]) {
>     return { sourceDescriptionName: parts[1], workflowId: parts[3], isLegacyForm: true };
>   }
> 
>   return undefined;
> }
> 
> export function resolveWorkflowReference({
>   ref,
>   ctx,
> }: {
>   ref: string | undefined;
>   ctx: TestContext | RuntimeExpressionContext;
> }): Workflow | undefined {
>   if (!ref) {
>     return undefined;
>   }
> 
>   if (!ref.startsWith('$')) {
>     return 'workflows' in ctx
>       ? ctx.workflows?.find((workflow) => workflow.workflowId === ref)
>       : undefined;
>   }
> 
>   const parsedRef = parseSourceDescriptionWorkflowRef(ref);
> 
>   if (!parsedRef) {
>     return undefined;
>   }
> 
>   const workflows = ctx.$sourceDescriptions?.[parsedRef.sourceDescriptionName]?.workflows;
> 
>   return Array.isArray(workflows)
>     ? workflows.find((workflow: Workflow) => workflow.workflowId === parsedRef.workflowId)
>     : undefined;
> }
> ```
> 
> then we could update our `get-value-from-context` condition to not return undefined, but throw errors when we have not valid refs
> 
> ```
> if (path.startsWith('$sourceDescriptions.')) {
>     const parsedRef = parseSourceDescriptionWorkflowRef(path);
> 
>     if (parsedRef) {
>       const { sourceDescriptionName, workflowId: reference, isLegacyForm } = parsedRef;
>       const sourceDescriptions = getFrom(ctx)('$sourceDescriptions');
>       const sourceDescriptionDocument = sourceDescriptions?.[sourceDescriptionName];
> 
>       if (!sourceDescriptionDocument) {
>         throw new Error(
>           `Can't resolve ${red(path)}: source description ${red(
>             sourceDescriptionName
>           )} is not found. Available source descriptions: ${Object.keys(
>             sourceDescriptions ?? {}
>           ).join(', ')}.`
>         );
>       }
> 
>       const workflow = resolveWorkflowReference({ ref: path, ctx });
> 
>       if (workflow) {
>         return workflow;
>       }
> 
>       if (!isLegacyForm) {
>         // Raw Source Description Objects (name, url, type) are only carried by the
>         // test context — deliberately not exposed on the runtime expression context.
>         const sourceDescriptionObject = (
>           'sourceDescriptions' in ctx ? ctx.sourceDescriptions : undefined
>         )?.find(({ name }) => name === sourceDescriptionName);
> 
>         if (
>           sourceDescriptionObject &&
>           Object.prototype.hasOwnProperty.call(sourceDescriptionObject, reference)
>         ) {
>           return (sourceDescriptionObject as Record<string, unknown>)[reference];
>         }
> 
>         if (Object.prototype.hasOwnProperty.call(sourceDescriptionDocument, reference)) {
>           return sourceDescriptionDocument[reference];
>         }
>       }
> 
>       const availableWorkflows = (
>         Array.isArray(sourceDescriptionDocument.workflows)
>           ? (sourceDescriptionDocument.workflows as Workflow[])
>           : []
>       )
>         .map((workflow) => workflow.workflowId)
>         .join(', ');
>       const availableWorkflowsHint = availableWorkflows
>         ? ` Available workflows: ${availableWorkflows}.`
>         : '';
> 
>       throw new Error(
>         isLegacyForm
>           ? `Can't resolve ${red(path)}: workflow ${red(
>               reference
>             )} is not found in source description ${red(
>               sourceDescriptionName
>             )}.${availableWorkflowsHint}`
>           : `Can't resolve ${red(path)}: ${red(
>               reference
>             )} does not match a workflow or a field of source description ${red(
>               sourceDescriptionName
>             )}.${availableWorkflowsHint}`
>       );
>     }
>     // Other shapes ($sourceDescriptions.<name>, deep paths) resolve generically below.
>   }
> ```
> 
> This will also require additional changes in the execution flow. run-step.ts
> 
> ```
> ....
>     const targetWorkflow = resolveWorkflowReference({ ref: targetWorkflowRef, ctx });
> ....
>         // register the step in catch sections so the failed check is displayed and counted in the totals
>         ctx.executedSteps.push(step);
>         printUnknownStep(step, ctx.options.logger);
> .....
> 
> and changes to the `runner.ts` as we need to register failed workflow when dependsOn not resolved properly:
> ```
> 
> .... try { await handleDependsOn({ workflow, ctx, config: options.config, executedStepsCount }); } catch (error) { if (!(error instanceof WorkflowDependencyNotFoundError)) { throw error; }
> 
> ```
>     // an unresolvable dependsOn reference fails this workflow but must not
>     // abort the remaining workflows and files
>     const startTime = performance.now();
>     const failedStep = {
>       stepId: 'dependsOn',
>       checks: [
>         {
>           name: CHECKS.UNEXPECTED_ERROR,
>           message: error.message,
>           passed: false,
>           severity: ctx.severity['UNEXPECTED_ERROR'],
>         },
>       ],
>     } as Step;
> 
>     printWorkflowSeparator({
>       fileName: basename(ctx.options.filePath),
>       workflowName: workflow.workflowId,
>       logger: options.logger,
>     });
>     printUnknownStep(failedStep, options.logger);
> 
>     executedWorkflows.push({
>       type: 'workflow',
>       workflowId: workflow.workflowId,
>       startTime,
>       endTime: performance.now(),
>       totalTimeMs: performance.now() - startTime,
>       executedSteps: [failedStep],
>       ctx,
>       globalTimeoutError: false,
>     });
>     continue;
>   }
> }
> ```
> ```
> 
> @harshit078 Could you please review this idea. Its raw implementation and will require additional test coverage, but might help to handle missing workflow referenses.

@harshit078 
I’ve implemented this proposed approach in the [PR](https://github.com/Redocly/redocly-cli/pull/2969). Feel free to review it and share any feedback.

**Comment by DmitryAnansky** [comment]
The ideas from this PR were incorporated into https://github.com/Redocly/redocly-cli/pull/2969.
@harshit078, thank you for your contribution and for taking the time to review it!
